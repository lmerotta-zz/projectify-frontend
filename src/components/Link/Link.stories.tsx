// Button.stories.tsx
import { ComponentProps } from "react";
import Link from "./Link";
import { Story } from "@storybook/react/types-6-0";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  title: "Link",
  component: Link,
};

const Template: Story<ComponentProps<typeof Link>> = (args) => (
  <Link {...args} href="#">
    My link
  </Link>
);

export const Default = Template.bind({});
