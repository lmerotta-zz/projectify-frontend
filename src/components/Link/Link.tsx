/** @jsxImportSource @emotion/react */
import { FC, HTMLProps } from "react";
import tw, { TwStyle } from "twin.macro";

type TextColor = "dark" | "secondary";

type LinkProps = {
  color?: TextColor;
};

const colorMapping: {
  [key in TextColor]: TwStyle;
} = {
  dark: tw`text-dark hover:text-default transition`,
  secondary: tw`text-secondary transition hover:text-secondary-hover`,
};

const Link: FC<
  Omit<HTMLProps<HTMLAnchorElement>, "as" | "color"> & LinkProps
> = ({ children, color = "dark", ...rest }) => {
  return (
    <a {...rest} css={colorMapping[color]}>
      {children}
    </a>
  );
};

export default Link;
