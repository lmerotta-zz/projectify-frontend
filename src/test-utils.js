import React from "react";
import { render } from "@testing-library/react";
import { CacheProvider } from "@emotion/react";
import createCache from "@emotion/cache";

const cache = createCache({ key: "iodajdoiwanoawdni" });
cache.compat = true;

const AllTheProviders = ({ children }) => {
  return <CacheProvider value={cache}>{children}</CacheProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
