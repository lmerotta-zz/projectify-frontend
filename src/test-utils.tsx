import { MockedProvider, MockedProviderProps } from "@apollo/client/testing";
import {
  CircularProgress,
  StyledEngineProvider,
  ThemeProvider,
} from "@mui/material";
import { render as realRender, RenderOptions } from "@testing-library/react";
import { ReactElement, ReactNode, Suspense } from "react";
import { MemoryRouter, MemoryRouterProps } from "react-router";
import { Route, Routes } from "react-router-dom";
import { ToastContainer } from "react-toastify";
import theme from "theme";

function renderWithProviders(
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries"> & {
    routerProps?: MemoryRouterProps;
    graphqlProps?: MockedProviderProps;
  }
) {
  const Wrapper = ({ children }: { children?: ReactNode }) => {
    return (
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={theme}>
          <ToastContainer
            theme="colored"
            autoClose={3000}
            containerId="react-toastify"
          />
          <MockedProvider {...options?.graphqlProps} addTypename={false}>
            <MemoryRouter {...options?.routerProps}>
              <Suspense fallback={<CircularProgress />}>
                <Routes>
                  <Route
                    path="/*"
                    element={
                      <>
                        <h1>TESTSASATATAST</h1>
                        {children}
                      </>
                    }
                  />
                </Routes>
              </Suspense>
            </MemoryRouter>
          </MockedProvider>
        </ThemeProvider>
      </StyledEngineProvider>
    );
  };

  return realRender(ui, { wrapper: Wrapper, ...options });
}

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
export { renderWithProviders };
