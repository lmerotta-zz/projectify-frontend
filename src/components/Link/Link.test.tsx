import { render } from "test-utils";
import Link from "./Link";
import each from "jest-each";
import { MemoryRouter } from "react-router";

describe("Link unit tests", () => {
  each([["dark"], ["secondary"]]).it("Renders with color %s", (color) => {
    const { asFragment } = render(
      <MemoryRouter>
        <Link to="#" color={color}>
          My Link
        </Link>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
