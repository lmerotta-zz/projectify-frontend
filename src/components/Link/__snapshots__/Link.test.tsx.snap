// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link unit tests Renders with color dark 1`] = `
<DocumentFragment>
  .emotion-0 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  --tw-text-opacity: 1;
  color: rgba(63, 63, 70, var(--tw-text-opacity));
}

.emotion-0:hover {
  --tw-text-opacity: 1;
  color: rgba(161, 161, 170, var(--tw-text-opacity));
}

<a
    class="emotion-0"
    href="#"
  >
    My Link
  </a>
</DocumentFragment>
`;

exports[`Link unit tests Renders with color secondary 1`] = `
<DocumentFragment>
  .emotion-0 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  --tw-text-opacity: 1;
  color: rgba(234, 179, 8, var(--tw-text-opacity));
}

.emotion-0:hover {
  --tw-text-opacity: 1;
  color: rgba(202, 138, 4, var(--tw-text-opacity));
}

<a
    class="emotion-0"
    href="#"
  >
    My Link
  </a>
</DocumentFragment>
`;
